src/domain/core/src/repos/metadata_chain_visitor.rs

```diff
@@ -7,73 +7,49 @@
-        const SET_INFO = 1 << 7;
-        const SET_LICENSE = 1 << 8;
-        const SET_DATA_SCHEMA = 1 << 9;
-        const ADD_PUSH_SOURCE = 1 << 10;
-        const DISABLE_PUSH_SOURCE = 1 << 11;
-        const DISABLE_POLLING_SOURCE = 1 << 12;
-        //
-        const DATA_BLOCK = Self::ADD_DATA.bits() | Self::EXECUTE_TRANSFORM.bits();
-    }
-}
-
-impl From<&MetadataBlock> for MetadataBlockTypeFlags {
-    fn from(block: &MetadataBlock) -> Self {
-        match block.event {
-            MetadataEvent::AddData(_) => Self::ADD_DATA,
-            MetadataEvent::ExecuteTransform(_) => Self::EXECUTE_TRANSFORM,
-            MetadataEvent::Seed(_) => Self::SEED,
-            MetadataEvent::SetPollingSource(_) => Self::SET_POLLING_SOURCE,
-            MetadataEvent::SetTransform(_) => Self::SET_TRANSFORM,
-            MetadataEvent::SetVocab(_) => Self::SET_VOCAB,
-            MetadataEvent::SetAttachments(_) => Self::SET_ATTACHMENTS,
-            MetadataEvent::SetInfo(_) => Self::SET_INFO,
-            MetadataEvent::SetLicense(_) => Self::SET_LICENSE,
-            MetadataEvent::SetDataSchema(_) => Self::SET_DATA_SCHEMA,
-            MetadataEvent::AddPushSource(_) => Self::ADD_PUSH_SOURCE,
-            MetadataEvent::DisablePushSource(_) => Self::DISABLE_PUSH_SOURCE,
-            MetadataEvent::DisablePollingSource(_) => Self::DISABLE_POLLING_SOURCE,
-        }
-    }
-}
-
-///////////////////////////////////////////////////////////////////////////////
-
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum Decision {
+pub enum MetadataVisitorDecision {
+    /// Stop Marker. A visitor who has reported the end of work will not be
+    /// invited to visit again
     Stop,
+    /// A request for the previous block
     Next,
+    /// A request for a specific previous block, specifying its hash.
+    /// Reserved for long jumps through the metadata chain
     NextWithHash(Multihash),
-    NextOfType(MetadataBlockTypeFlags),
+    /// A request for previous blocks, of specific types, using flags.
+    ///
+    /// # Examples
+    /// ```
+    /// // Request for SetVocab block
+    /// return MetadataVisitorDecision::NextOfType(MetadataEventTypeFlags::SET_VOCAB);
+    ///
+    /// // Request for data blocks (ADD_DATA || EXECUTE_TRANSFORM)
+    /// return MetadataVisitorDecision::NextOfType(MetadataEventTypeFlags::DATA_BLOCK);
+    ///
+    /// // Request for a list of blocks of different types
+    /// return MetadataVisitorDecision::NextOfType(MetadataEventTypeFlags::SET_ATTACHMENTS | MetadataEventTypeFlags::SET_LICENSE);
+    /// ```
+    NextOfType(MetadataEventTypeFlags),
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
 pub trait MetadataChainVisitor: Sync + Send {
     type Error: std::error::Error;
 
-    fn visit(&mut self, hashed_block_ref: HashedMetadataBlockRef) -> Result<Decision, Self::Error>;
+    fn initial_decision(&self) -> Result<MetadataVisitorDecision, Self::Error>;
```

Comment:
> Under what circumstances would `initial_decision()` return an error?

I originally thought this might serve as a simplification -- it worked for part of the cases, but yes, the examples you give show that it may not be the case

--

Now `initial_decision()` returns onlu `MetadataVisitorDecision`